### PR TITLE
Standardize backend response

### PR DESCRIPTION
**Pull request by dsr5964**
I am torn on how to do errors. I was considering adding an error structure that would appear if there is an error and it won't if the response was successful. You can let me know what you think. There are probably some things I'm missing in the response which I can add but I want you to take a look before the meeting. I initially had the GetResponse split into two separate functions one that included pagination in the response and one that did not. I decided it would be best to pass the page and limit parameters as pointers and if the pointers were not nil then include pagination in the response. If you are confused what I'm talking about you can look at the previous commits. I didn't include an error return in GetResponse because then we would have to do another error check if the controller method, you can let me know what you think if it is necessary or not. I have no issues including it I just think it gets a bit repetitive. I tested this it works I also tested the edge cases as well if you were wondering.

### CODE REVIEW

File: backend/models/metadata.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package models
+
+type Meta struct {
```
Comment (phill52) [inline backend/models/metadata.go:3]: The models package is just for database models, otherwise it will get confusing what's a DB model and what's just a struct we're using and it'll get too big. Should probably make a utilities package, and response module in it.

File: backend/repositories/response.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+func GetResponse(model interface{}, page, limit *int, statusCode int, message string) *models.Response {
+	meta := GetMetaData(message, statusCode)
+
+	response := &models.Response{
+		Data: model,
+		Meta: *meta,
+	}
+
+	if page != nil && limit != nil {
+		query := database.DB.Model(&model)
```
Comment (phill52) [inline backend/repositories/response.go:18]: I don't like how this is structured, this function shouldn't be doing any of the business logic, it should just take some inputs and return a standardized response. Maybe make it so the pagination data is optional in the metadata, there is a pagination data arg in the function that if not nil will add the pagination data to the metadata. Idk if this'll work, but worth a try. Ill think about it more.

File: backend/repositories/response.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package repositories
+
+import (
+	"smashfriend/database"
+	"smashfriend/models"
+	"smashfriend/utils"
+)
+
```
Comment (phill52) [inline backend/repositories/response.go:8]: Repositories should be pure just database functions, like I said earlier, put this in a utility package.

File: backend/main.go [hunk 4]
```diff
@@ -27,7 +27,6 @@ func main() {
 	router.GET("/users", controllers.GetUsers)
 	router.GET("/users/:id", controllers.GetUser)
 	router.POST("/users", controllers.CreateUser)
-
```
Comment (phill52) [inline backend/main.go:30]: Keep this blank line, we'll eventually have a lot of endpoints so we'll have to group them up.

File: backend/controllers/users.go [hunk 5]
```diff
@@ -17,63 +17,76 @@ func GetUsers(c *gin.Context) {
...
 	if err != nil {
-		c.JSON(http.StatusBadRequest, gin.H{"error": "page parameter not valid"})
+		response := repositories.GetResponse(nil, nil, nil, http.StatusBadRequest, "Bad Request: page parameter not valid")
+		c.JSON(http.StatusBadRequest, response)
 		return
 	}
 	limit, err := strconv.Atoi(limitStr)
 	if err != nil {
-		c.JSON(http.StatusBadRequest, gin.H{"error": "limit parameter not valid"})
+		response := repositories.GetResponse(nil, nil, nil, http.StatusBadRequest, "Bad Request: limit parameter not valid")
```
Comment (phill52) [inline backend/controllers/users.go:26]: Don't put "bad response:" in the the message, that's already implied by the error code. The whole message is going to be displayed by the frontend to the user, so it should just be a statement of what the error condition is, I think some of your other ones got it good.

File: backend/controllers/users.go [hunk 6]
```diff
@@ -17,63 +17,76 @@ func GetUsers(c *gin.Context) {
...
+	response := repositories.GetResponse(user, nil, nil, http.StatusOK, "Success")
+	c.JSON(http.StatusOK, response)
 }
 
 func CreateUser(c *gin.Context) {
 	username := c.PostForm("username")
 
 	if len(username) < 3 {
-		c.JSON(http.StatusBadRequest, gin.H{"error": "Username must be at least 3 characters long"})
+		response := repositories.GetResponse(nil, nil, nil, http.StatusBadRequest, "Username must be at least 3 characters long")
```
Comment (phill52) [inline backend/controllers/users.go:67]: Yeah this one is good

File: backend/controllers/users.go (hunk 5 above)
Comment (dsr5964) [reply backend/controllers/users.go:26]: I had that earlier where I put our own message "page parameter is not valid" if the strconv.Atoi() is not valid. So should I change it back?

File: backend/repositories/response.go (hunk 2 above)
Comment (dsr5964) [reply backend/repositories/response.go:18]: So you want the pagination to be within the metadata structure as opposed to within the response structure?

Comment (dsr5964) [reply backend/repositories/response.go:18]: The reason I passed in the page and limit parameters into this method primarily was because the GetResponse is being called in the controller. The pagination is being called within the repository of GetUsers and I didn't want to change it so GetUsers returns a response because it probably should return []User. There might be a way to do it but there is no way I can pass the Pagination into GetResponse without changing how GetUser. What I can try is including the Pagination within MetaData and calling GetMetaData in the repository of GetUser.

Comment (dsr5964) [reply backend/repositories/response.go:18]: Yeah I realized I can't call GetMetaData because I would need to the status code and response.

File: backend/utils/response.go [hunk 7]
```diff
@@ -1,41 +1,35 @@
...
 	"smashfriend/utils/response"
 )
 
-func GetResponse(model interface{}, pagination interface{}, page, limit *int, statusCode int, message string) *response.Response {
-	meta, err := GetMetaData(model, pagination, page, limit, message, statusCode)
-	if err != nil {
-		response := response.Response{}
+func GetResponse(data, pagination interface{}, statusCode int, message string) *response.Response {
+	meta := GetMetaData(pagination, statusCode, message)
+	if meta.Pagination == nil && statusCode != 200 {
```
Comment (dsr5964) [inline backend/utils/response.go:9]: I am not a fan of this approach but if we don't check if the statusCode is not 200 then it will return an error for GetUserById and CreateUser as well. We basically are saying if the PaginationStructure passed into the function is nil then return an error structure for GetPaginatedUsers. Since GetUserById and CreateUser will not have pagination we need to check the status code. I'm not sure if there is a better way to get around this.

**Review by phill52 — DISMISSED** [review]
This looks good to merge. There are a couple things I'll comment that are worth noting for future reference, but aren't worth blocking this for. Let's finally get started on the meat of the API.

File: backend/repositories/users.go [hunk 8]
```diff
@@ -8,17 +8,27 @@ import (
 	"smashfriend/utils"
 )
 
-func GetUsers(page, limit int) ([]models.User, error) {
+type PaginatedUsers struct {
```
Comment (phill52) [inline backend/repositories/users.go:11]: We should probably turn this to a more generic type, because there are going to be other paginated routes. We can do this when we get to those though.

File: backend/controllers/users.go [hunk 9]
```diff
@@ -17,63 +17,78 @@ func GetUsers(c *gin.Context) {
...
+			response := utils.GetResponse(nil, nil, http.StatusInternalServerError, "")
+			c.JSON(http.StatusInternalServerError, response)
 			return
 		}
 	}
 
-	c.JSON(http.StatusOK, users)
+	users := paginatedUsers.Users
+	paginationData := paginatedUsers.Pagination
+	response := utils.GetResponse(users, paginationData, http.StatusOK, "Success")
```
Comment (phill52) [inline backend/controllers/users.go:47]: Making the users and paginationData variables before this aren't really necessary because we aren't using them anywhere else, and it just adds extra lines. You could just do `GetResponse(paginatedUsers.Users, pagiantedUsers.pagination ...`. 
You generally don't want your code growing vertically, because it becomes more annoying to navigate. The cases you would make a single use variable is:
1. the function call line is getting too long (> 125 characters)
2. there starts being nested parentheses that make it too complicated to read ie. `GetUser(GetUserID(ParseToken(request.Header)))`
3. what the object is, or the intent of it, is not immediately clear, and a variable name could help.

**Comment by phill52** [comment]
Wait actually, now that I look at it again, I think it might be simpler to have getresponse and geterror be more divorced. When you're in a controller and you want to return an error, you know in that line that you want to, so you can just call a `GetErrorResponse` function that takes the error code and message, and when it's a success, you can do the standard `GetResponse` function. The logic in `GetResponse` just looks a little more complicated than it needs to be. 
If this can be done simply, try that, but if there's more folds to it, you can just merge this for now.

**Review by phill52 — APPROVED** [review]
